src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/Orleans3CompatibleStringKeyHasher.cs

```diff
@@ -37,17 +38,63 @@ public int Hash(byte[] data)
             // PickHasher parameters are the same for both calls so we need to analyze data content to distinguish these cases.
             // It doesn't word if string key is equal to grain type name, but we consider this edge case to be negligibly rare.
 
-            // reducing allocations if data is not a grain type
-            if (data.Length >= _grainType.Length && Encoding.UTF8.GetByteCount(_grainType) == data.Length)
+            if (IsGrainTypeName(data))
+                return _innerHasher.Hash(data);
+
+            var extendedLength = data.Length + 8;
+            if (extendedLength <= 256)
+            {
+                Span<byte> extended = stackalloc byte[extendedLength];
+                data.AsSpan().CopyTo(extended);
+                return _innerHasher.Hash(extended);
+            }
+
+            var buffer = ArrayPool<byte>.Shared.Rent(extendedLength);
+            try
+            {
+                data.AsSpan().CopyTo(buffer);
+                Array.Clear(buffer, data.Length, 8);
```

Comment:
Is the `Clear` needed?
For the stack-alloc path above you can safely assume that the stack-space is zeroed. E.g. when there's `[SkipLocalsInit]` applied.

I'd combine the two pathes, to avoid the duplicated logic (though it's very little logic here).

Reply:
yep, we need to clear rented buffer to make sure all these bytes set to zero - rented buffer can contain arbitrary data

Reply:
Ah yes, I misread the code here. The extra 8 bytes need to be zeroed.
Thanks for the hint!

As the span also needs to be cleared, these two code-pathes should be collapsed (see other comments).